File: Estruturais/Flyweight/flyweight.py
# flyweight  factory ------------------------------
class FlyweightFactory:
    def __init__(self):
        self.flyweights = {}

    def getFlyweight(self, key):
        if key not in self.flyweight:
            self.flyweights[key] = Flyweight(key)
        return self.flyweights[key]        

# flyweight ------------------------------------

class Flyweight:
    def __init__(self, intrinsicState):
        self.intrinsicState = intrinsicState

    def operation(self, extrinsicState):
        print(f'Intrinsico: {self.intrinsicState}, Extrinsico: {self.extrinsicState}')         


# cliente -----------------------------------------
class Client:
    def __init__(self):
        self.flyweightFactory = FlyweightFactory()

    def run(self):
        flyweightA = self.flyweightFactory.getFlyweight('A')
        flyweightB = self.flyweightFactory.getFlyweight('B')
        flyweightC = self.flyweightFactory.getFlyweight('C')

        flyweightA.operation('Estado 1')
        flyweightB.operation('Estado 2')
        flyweightC.operation('Estado 3')

# usase -----------------------------------------
client = Client()
client.run()

File: Estruturais/Flyweight/flyweight.js
// flyweight  factory ---------------------------------------------------------------------------
class FlyweightFactory {
    constructor() {
        this.flyweights = {}
    }

    getFlyweight(key) {
        if(!this.flyweights[key]) {
            this.flyweights[key] = new Flyweight(key);
        }
        return this.flyweights[key];
    }
}
//flyweight ---------------------------------------------------------------------------
class Flyweight {
    constructor(intrinsicState){
        this.intrinsicState = intrinsicState;
    }

    operation(extrinsicState) {
        console.log(`Intrinsico: ${this.intrinsicState}, Extrinsico: ${extrinsicState}`);
        //intrinsico objs tem em comum. Extrinsico objs tem de diferente
    }
}

//cliente ---------------------------------------------------------------------------
class Client {
    constructor(){
        this.flyweightFactory = new FlyweightFactory();
        
    }

    run (){
        const flyweightA = this.flyweightFactory.getFlyweight('A');
        const flyweightB = this.flyweightFactory.getFlyweight('B');
        const flyweightC = this.flyweightFactory.getFlyweight('C');

        flyweightA.operation('Estado 1');
        flyweightB.operation('Estado 2');
        flyweightC.operation('Estado 3');
    }

}
// usase
const client = new Client();
client.run();